content/post/ddos-detection.md

```diff
@@ -0,0 +1,413 @@
+
+```python
+def update_current(self, data):
+    self.current_req += 1
+    self.current_cli.add(data['client'])
+```
+
+While performing the status check requires some care in handling the 4 transitions:
+
+1. healthy --> under_attack
+2. healthy --> healthy
+3. under_attack --> healthy
+4. under_attack --> under_attack
+
+As well as in adhering to the minimum thresholds required for a status change
+
+```python
+def check_status(self, timestamp):
+    ts_frac = timestamp % 1
+    if ts_frac > 0 and self.current_req > 10:
+        # compute the expected final value for the current aggregator
+        exp_cli = len(self.current_cli)/ts_frac
+        exp_req = self.current_req/ts_frac
+
+        # is it >2sigma from the mean?
+        is_attack = False
+        if (exp_cli - self.cli_mu) > (2 * self.cli_sigma):
+            is_attack = True
+        elif (exp_req - self.req_mu) > (2 * self.cli_sigma):
+            is_attack = True
+
+        if is_attack:
+            if self.is_attack:
+                # under_attack --> under_attack
+                return None
+            else:
+                if self.current_req > 20:
+                    # healthy --> under_attack
+                    self.is_attack = True
+                    return (self.address, timestamp, True)
+        else:
+            if self.is_attack:
+                # under_attack --> healthy
+                self.is_attack = False
+                return (self.address, timestamp, False)
+            else:
+                # healthy --> healthy
+                return None
+```
+
+However, we are partitioning into IP range segments, and not unique IPs, so we still need a state object per partition, along with its builder
+
+```python
+class PartitionStatsBuilder(object):
+    def build(self):
+        return PartitionLoadStatistics()
+
+
+class PartitionLoadStatistics(object):
+    def __init__(self):
+        self.servers = {}
+
+    def update(self, data):
+        return (self.servers.setdefault(data['server'],
+                                        SingleServerLoadStatistics(
+                                            data['server']))
+                .update(data))
+```
+
+Lastly, we still need to encode the output data for a TCP consumer
+
+```python
+class Encoder(object):
+    def encode(self, data):
+        # data is a tuple of (server_name, timestamp, is_under_attack)
+        if data[2]:
+            # under attack
+            return ("Server {} is under ATTACK! (Status changed at {})\n"
+                    .format(data[0], data[1]))
+        else:
+            # no longer under attack
+            return ("Server {} is no longer under attack. "
+                    "(Status changed at {})\n"
+                    .format(data[0], data[1]))
+```
+
+And finally, perform the imports, define our application's serialisation methods, and set up the application and its pipeline so Wallaroo can run it all:
```

Comment:
In the sentence above you said "Lastly ...", and in this sentence you begin with "And finally ...". I'd drop one of those, since they mean more or less the same thing and you can't have two last/final steps.